Sources/AsyncHTTPClient/AsyncAwait/Transaction+StateMachine.swift

```diff
@@ -0,0 +1,603 @@
+                assert(!buffer.isEmpty)
+                let toReturn = buffer.removeFirst()
+                self.state = .executing(context, requestState, .buffering(streamID, buffer, next: .endOfFile))
+                return .succeedContinuation(continuation, toReturn)
+
+            case .executing(let context, let requestState, .buffering(let registeredStreamID, var buffer, next: .askExecutorForMore)):
+                guard registeredStreamID == streamID else {
+                    return .failContinuation(continuation, TriedToRegisteredASecondConsumer())
+                }
+
+                if buffer.isEmpty {
+                    self.state = .executing(context, requestState, .waitingForRemote(streamID, continuation))
+                    return .askExecutorForMore(context.executor)
+                } else {
+                    let toReturn = buffer.removeFirst()
+                    self.state = .executing(context, requestState, .buffering(streamID, buffer, next: .askExecutorForMore))
+                    return .succeedContinuation(continuation, toReturn)
+                }
+
+            case .executing(_, _, .buffering(let registeredStreamID, _, next: .error(let error))):
+                guard registeredStreamID == streamID else {
+                    return .failContinuation(continuation, TriedToRegisteredASecondConsumer())
+                }
+                self.state = .finished(error: error, registeredStreamID)
+                return .failContinuation(continuation, error)
+
+            case .executing(_, _, .buffering(let registeredStreamID, let buffer, next: .endOfFile)) where buffer.isEmpty:
+                guard registeredStreamID == streamID else {
+                    return .failContinuation(continuation, TriedToRegisteredASecondConsumer())
+                }
+                self.state = .finished(error: nil, registeredStreamID)
+                return .succeedContinuation(continuation, nil)
+
+            case .executing(let context, let requestState, .buffering(let registeredStreamID, var buffer, next: .endOfFile)):
+                guard registeredStreamID == streamID else {
+                    return .failContinuation(continuation, TriedToRegisteredASecondConsumer())
+                }
+                assert(!buffer.isEmpty) // TODO: This looks fishy!
+                let toReturn = buffer.removeFirst()
+                self.state = .executing(context, requestState, .buffering(streamID, buffer, next: .endOfFile))
+                return .succeedContinuation(continuation, toReturn)
+
+            case .executing(_, _, .waitingForRemote(let registeredStreamID, _)):
+                guard registeredStreamID == streamID else {
+                    return .failContinuation(continuation, TriedToRegisteredASecondConsumer())
+                }
+                preconditionFailure("A body response continuation from this iterator already exists! Queuing calls to `next()` is not supported.")
```

Comment:
Is there any reason this `preconditionFailure` is less important than the check above?